File: backend-java/src/main/java/com/wealthtech/crm/modules/usermanager/service/PermissionSeeder.java
package com.wealthtech.crm.modules.usermanager.service;

import com.wealthtech.crm.modules.usermanager.entity.Permission;
import com.wealthtech.crm.modules.usermanager.repository.PermissionRepository;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Component;

@Component
public class PermissionSeeder implements CommandLineRunner {
    private final PermissionRepository permissionRepository;
    private final PermissionService permissionService;

    public PermissionSeeder(PermissionRepository permissionRepository, PermissionService permissionService) {
        this.permissionRepository = permissionRepository;
        this.permissionService = permissionService;
    }

    @Override
    public void run(String... args) {
        if (permissionRepository.count() > 0) {
            return;
        }

        // Clients
        seedPermission("client:read", "View Client", "client");
        seedPermission("client:create", "Create Client", "client");
        seedPermission("client:update", "Update Client", "client");
        seedPermission("client:delete", "Delete Client", "client");

        // Client Profiles
        seedPermission("clientprofile:read", "View Client Profile", "clientprofile");
        seedPermission("clientprofile:create", "Create Client Profile", "clientprofile");
        seedPermission("clientprofile:update", "Update Client Profile", "clientprofile");
        seedPermission("clientprofile:delete", "Delete Client Profile", "clientprofile");

        // Risk Appetite
        seedPermission("riskappetite:read", "View Risk Appetite", "riskappetite");
        seedPermission("riskappetite:create", "Create Risk Appetite", "riskappetite");
        seedPermission("riskappetite:update", "Update Risk Appetite", "riskappetite");
        seedPermission("riskappetite:delete", "Delete Risk Appetite", "riskappetite");

        // Portfolio Review & Entry
        seedPermission("portfolioreview:read", "View Portfolio Review", "portfolioreview");
        seedPermission("portfolioreview:create", "Create Portfolio Review", "portfolioreview");
        seedPermission("portfolioreview:update", "Update Portfolio Review", "portfolioreview");
        seedPermission("portfolioreview:delete", "Delete Portfolio Review", "portfolioreview");

        seedPermission("portfolioentry:read", "View Portfolio Entry", "portfolioentry");
        seedPermission("portfolioentry:create", "Create Portfolio Entry", "portfolioentry");
        seedPermission("portfolioentry:update", "Update Portfolio Entry", "portfolioentry");
        seedPermission("portfolioentry:delete", "Delete Portfolio Entry", "portfolioentry");

        // Tasks & Notes
        seedPermission("task:read", "View Task", "task");
        seedPermission("task:create", "Create Task", "task");
        seedPermission("task:update", "Update Task", "task");
        seedPermission("task:delete", "Delete Task", "task");

        seedPermission("note:read", "View Note", "note");
        seedPermission("note:create", "Create Note", "note");
        seedPermission("note:update", "Update Note", "note");
        seedPermission("note:delete", "Delete Note", "note");

        // User Management: Users, Groups, Roles
        seedPermission("user:read", "View User", "user");
        seedPermission("user:create", "Create User", "user");
        seedPermission("user:update", "Update User", "user");
        seedPermission("user:delete", "Delete User", "user");

        seedPermission("userprofile:read", "View User Profile", "userprofile");

        seedPermission("crmgroup:read", "View CRM Group", "crmgroup");
        seedPermission("crmgroup:create", "Create CRM Group", "crmgroup");
        seedPermission("crmgroup:update", "Update CRM Group", "crmgroup");
        seedPermission("crmgroup:delete", "Delete CRM Group", "crmgroup");

        seedPermission("role:read", "View Role", "role");
        seedPermission("role:create", "Create Role", "role");
        seedPermission("role:update", "Update Role", "role");
        seedPermission("role:delete", "Delete Role", "role");

        seedPermission("rolepermission:read", "View Role Permission", "rolepermission");
        seedPermission("rolepermission:create", "Create Role Permission", "rolepermission");
        seedPermission("rolepermission:update", "Update Role Permission", "rolepermission");
        seedPermission("rolepermission:delete", "Delete Role Permission", "rolepermission");
    }

    private void seedPermission(String name, String displayName, String resource) {
        permissionService.getOrCreate(name, displayName, resource);
    }
}

File: backend-nodejs/src/modules/usermanager/seeders/permissionSeeder.ts
import { Permission } from '../models/Permission';
import { logger } from '../../../common/utils/logger';

const PERMISSIONS_DATA = [
  // Clients
  { codename: 'client:read', name: 'View Client', contentType: 'client' },
  { codename: 'client:create', name: 'Create Client', contentType: 'client' },
  { codename: 'client:update', name: 'Update Client', contentType: 'client' },
  { codename: 'client:delete', name: 'Delete Client', contentType: 'client' },

  // Client Profiles
  { codename: 'clientprofile:read', name: 'View Client Profile', contentType: 'clientprofile' },
  { codename: 'clientprofile:create', name: 'Create Client Profile', contentType: 'clientprofile' },
  { codename: 'clientprofile:update', name: 'Update Client Profile', contentType: 'clientprofile' },
  { codename: 'clientprofile:delete', name: 'Delete Client Profile', contentType: 'clientprofile' },

  // Risk Appetite
  { codename: 'riskappetite:read', name: 'View Risk Appetite', contentType: 'riskappetite' },
  { codename: 'riskappetite:create', name: 'Create Risk Appetite', contentType: 'riskappetite' },
  { codename: 'riskappetite:update', name: 'Update Risk Appetite', contentType: 'riskappetite' },
  { codename: 'riskappetite:delete', name: 'Delete Risk Appetite', contentType: 'riskappetite' },

  // Portfolio Review & Entry
  { codename: 'portfolioreview:read', name: 'View Portfolio Review', contentType: 'portfolioreview' },
  { codename: 'portfolioreview:create', name: 'Create Portfolio Review', contentType: 'portfolioreview' },
  { codename: 'portfolioreview:update', name: 'Update Portfolio Review', contentType: 'portfolioreview' },
  { codename: 'portfolioreview:delete', name: 'Delete Portfolio Review', contentType: 'portfolioreview' },

  { codename: 'portfolioentry:read', name: 'View Portfolio Entry', contentType: 'portfolioentry' },
  { codename: 'portfolioentry:create', name: 'Create Portfolio Entry', contentType: 'portfolioentry' },
  { codename: 'portfolioentry:update', name: 'Update Portfolio Entry', contentType: 'portfolioentry' },
  { codename: 'portfolioentry:delete', name: 'Delete Portfolio Entry', contentType: 'portfolioentry' },

  // Tasks & Notes
  { codename: 'task:read', name: 'View Task', contentType: 'task' },
  { codename: 'task:create', name: 'Create Task', contentType: 'task' },
  { codename: 'task:update', name: 'Update Task', contentType: 'task' },
  { codename: 'task:delete', name: 'Delete Task', contentType: 'task' },

  { codename: 'note:read', name: 'View Note', contentType: 'note' },
  { codename: 'note:create', name: 'Create Note', contentType: 'note' },
  { codename: 'note:update', name: 'Update Note', contentType: 'note' },
  { codename: 'note:delete', name: 'Delete Note', contentType: 'note' },

  // User Management
  { codename: 'user:read', name: 'View User', contentType: 'user' },
  { codename: 'user:create', name: 'Create User', contentType: 'user' },
  { codename: 'user:update', name: 'Update User', contentType: 'user' },
  { codename: 'user:delete', name: 'Delete User', contentType: 'user' },

  { codename: 'userprofile:read', name: 'View User Profile', contentType: 'userprofile' },

  { codename: 'crmgroup:read', name: 'View CRM Group', contentType: 'crmgroup' },
  { codename: 'crmgroup:create', name: 'Create CRM Group', contentType: 'crmgroup' },
  { codename: 'crmgroup:update', name: 'Update CRM Group', contentType: 'crmgroup' },
  { codename: 'crmgroup:delete', name: 'Delete CRM Group', contentType: 'crmgroup' },

  { codename: 'role:read', name: 'View Role', contentType: 'role' },
  { codename: 'role:create', name: 'Create Role', contentType: 'role' },
  { codename: 'role:update', name: 'Update Role', contentType: 'role' },
  { codename: 'role:delete', name: 'Delete Role', contentType: 'role' },

  { codename: 'rolepermission:read', name: 'View Role Permission', contentType: 'rolepermission' },
  { codename: 'rolepermission:create', name: 'Create Role Permission', contentType: 'rolepermission' },
  { codename: 'rolepermission:update', name: 'Update Role Permission', contentType: 'rolepermission' },
  { codename: 'rolepermission:delete', name: 'Delete Role Permission', contentType: 'rolepermission' },
];

export const seedPermissions = async (): Promise<void> => {
  try {
    const count = await Permission.countDocuments();
    if (count > 0) {
      return;
    }

    await Permission.insertMany(PERMISSIONS_DATA);
    logger.info({ count: PERMISSIONS_DATA.length }, 'Permissions catalogue seeded successfully');
  } catch (error: unknown) {
    const err = error instanceof Error ? error.message : String(error);
    logger.error({ error: err }, 'Failed to seed permissions catalogue');
  }
};
